kube-client/src/config/mod.rs

```diff
@@ -206,13 +208,32 @@ impl Config {
         Ok(config)
     }
 
-    /// Create configuration from the cluster's environment variables
+    /// Load an in-cluster config using the `KUBERNETES_SERVICE_HOST` and
+    /// `KUBERNETES_SERVICE_PORT` environment variables.
     ///
-    /// This follows the standard [API Access from a Pod](https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster/#accessing-the-api-from-a-pod)
-    /// and relies on you having the service account's token mounted,
-    /// as well as having given the service account rbac access to do what you need.
+    /// This matches the behavior of the official Kubernetes client libraries,
+    /// but it may not be compatible with `rustls`.
+    ///
+    /// A service account's token must be available in
+    /// `/var/run/secrets/kubernetes.io/serviceaccount/`.
     pub fn from_cluster_env() -> Result<Self, InClusterError> {
-        let cluster_url = incluster_config::kube_dns();
+        let uri = incluster_config::try_kube_from_env()?;
+        Self::load_inluster_with_uri(uri)
+    }
+
+    /// Load an in-cluster config using the API server at
+    /// `https://kubernetes.default.svc`.
+    ///
+    /// This behavior does not match that of the official Kubernetes clients,
+    /// but this approach is compatible with `rustls`.
+    ///
+    /// A service account's token must be available in
+    /// `/var/run/secrets/kubernetes.io/serviceaccount/`.
+    pub fn from_cluster_dns() -> Result<Self, InClusterError> {
```

Comment:
Alternatively, we could feature-gate the behavior based on `rustls-tls` so that `Config::infer` "does the right thing." This would let applications work with rustls in _many more_ cases. But people may run into problems like linkerd/linkerd2#9339. AND/OR we could eagerly panic if the `rustls-tls` feature is used with the environment configuration (because it's almost guaranteed to fail), but we'd at least have an opportunity to provide a better error message.